Openssl 3.0

ifranzki opened the pull request:
A series of patches that remove any OpenSSL functions that are deprecated with OpenSSL 3.0.
At some places conditional compile is used to have OpenSSL 1.1.1 specific code and OpenSSL 30 specific code separated, other places are changed so that it works for both versions.

While at it, remove support for OpenSSL versions older than v1.1.1. 

This addresses issue https://github.com/opencryptoki/opencryptoki/issues/415 .

Please note that although deprecated functions are supposed to continue to function as before, this is not entirely true for OpenSSL 3.0.  The Soft token uses functions `EVP_MD_meth_get_app_datasize()` and `EVP_MD_CTX_md_data()` to get and set the digests state in-between C_DigestInit, C_DigestUpdate, C_DigestFiinal. This works with OpenSSL 1.1.1 (although its somewhat of a hack...), but it no longer works with OpenSSL 3.0. `EVP_MD_meth_get_app_datasize()` always returns 0 which leads to the fact that no digest state is saved at all. Thus, any multipart digest operation will produce wrong output. 

This part of the code is reworked so that it works with OpenSSL 3.0, however, it is not possible with OpenSSL 3.0 to export the digest state in a way so that it could be obtained via `C_GetOperationState()`.
A token must therefore place pointers to OpenSSL digest contexts into the session state structure.
Such a state can not be externalized through `C_GetOperationState()`. `CKR_STATE_UNSAVEABLE` is returned by `C_GetOperationState()`. if such unsavable state is used by a token.

juergenchrist commented on testcases/crypto/rsa_func.c:
```diff
@@ -148,6 +148,16 @@ CK_RV do_EncryptDecryptRSA(struct GENERATED_TEST_SUITE_INFO *tsuite)
+        if (is_soft_token(slot_id)) {
+            if (!is_valid_soft_pubexp(tsuite->tv[i].publ_exp,
+                                      tsuite->tv[i].publ_exp_len)) {
+                testcase_skip("Soft Token cannot "
```
Ugly line wrap.  Makes it impossible to grep for that skip message.  Why don't we leave the string intact and put the argument into a new line.  Makes a better usability and still okay readability.

ifranzki replied:
will fix

juergenchrist commented on usr/lib/api/api_interface.c:
```diff
@@ -37,6 +37,28 @@
+        } while (0);
+#else
+#define BEGIN_OPENSSL_LIBCTX(ossl_ctx, rc)
+#define END_OPENSSL_LIBCTX(rc)
```
To enforce consistent usage between OpenSSL 1.1.1 and 3.0 I would define these two macros also as start and end of a scope (either `{` and `}` or `do {` and `} while (0);`.  Makes it easier to spot compile errors later one if one person changes code with 1.1.1 and that code does not compile due to wrong scoping on 3.0.

ifranzki replied:
good point, will change.

juergenchrist commented on configure.ac:
```diff
@@ -39,10 +39,8 @@ dnl Checks for library functions.
-AC_FUNC_MALLOC
 AC_FUNC_MKTIME
 AC_FUNC_MMAP
-AC_FUNC_REALLOC
```
Those two checks basically make sure that allocating 0 bytes yields a valid pointer.  I am not aware of any 0-byte allocations.  So I think we don't need them.  Otherwise we should use gnulibs `malloc-gnu` and `realloc-gnu` modules.

ifranzki replied:
> I am not aware of any 0-byte allocations

right. 
I would rather like that a malloc(0) fails with a NULL pointer, that way you spot such coding errors more easily, at least at runtime.

juergenchrist reviewed (commented):
Since all OpenSSL pre-1.1.1 support is dropped, shouldn't we implement a configure check to make sure we are not compiling against such a version?  Would probably be nicer than just giving random compile errors.  Also, does it run with OpenSSL 1.1.0?  The SONAME of OpenSSL only covers major and minor, but our version number check always checked for 1.1.1.

ifranzki commented on usr/lib/common/utility.c:
```diff
@@ -849,66 +849,89 @@ CK_RV get_hmac_digest(CK_ULONG mech, CK_ULONG *digest_mech, CK_BBOOL *general)
-    if (rv != CKR_OK) {
-        TRACE_DEBUG("failed to create digest.\n");
-        return rv;
+    if (EVP_Digest(data, len, hash, &hash_len, md, NULL) != 1) {
```
Did you already take a look at commit https://github.com/opencryptoki/opencryptoki/pull/416/commits/22db88ca2a086c8824326765ba738b4b5a8193ae ? This should solve such loops.

juergenchrist replied:
That takes care of OpenSSL 3.0.  What about 1.1.1?  There, we just made the problem even bigger.  Since you changed common code, are now any tokens impacted that were not impacted before?

ifranzki replied:
As far as I know even before all tokens have been affected by that problem anyway. The private token encryption routines (sw_des3_cbc in sw_crypt.c) are already using the EVP interface since longer time. So this would have caused the loop already for all tokens anyway. 

Based on this, I would say the changes of this PR don't make the situation worse (but also not better for OpenSSL 1.1.1). 

I also don't know how one can force OpenSSL to NOT use a configured engine (if any). Passing NULL as engine to those functions that accept an engine doesn't help, NULL just means use whatever is configured.

ifranzki replied:
Maybe we could iterate over all registered engines via `ENGINE_get_first()`/`ENGINE_get_next()`, and remove all of them via `ENGINE_remove()` on entry to OCK code, and re-add them via `ENGINE_add()` when OCK returns back to the application? But this sounds really like a huge overhead.....

ifranzki commented on usr/lib/common/defs.h:
```diff
@@ -17,6 +17,20 @@
+            OPENSSL_VERSION_MINOR >= ((maj) << 16) + (min))
+    #else
+        #define OPENSSL_VERSION_PREREQ(maj, min)        \
+            (OPENSSL_VERSION_NUMBER >= (((maj) << 28) | \
```
Well, then check is if OPENSSL_VERSION_NUMBER is larger or equal to the version number built from may and min. So if OPENSSL_VERSION_NUMBER includes a pre-relaese in the 4 lowest bits, then it would still be larger or equal to the version built from may and min, given that the upper bits parts match. 
If you remove (i.e. zeroize) the 4 lower bits, then it would not make a big difference, any pre-release would still match.

juergenchrist replied:
Well, "remove" was a bad choice of words.  What I meant is "should a pre-release match?"

ifranzki replied:
> should a pre-release match?

I would say yes. The whole block is only used for pre OpenSSL 3.0 anyway, because in OpenSSL 3.0 OPENSSL_VERSION_PREREQ is already defined.
So if there ever is a pre-release version for 1.1.1 then I guess I would still be able to compile against it. It will anyway match 1.1.x, because it only checks major and minor. So I guess your question does not really apply :-)

ifranzki commented:
> Since all OpenSSL pre-1.1.1 support is dropped, shouldn't we implement a configure check to make sure we are not compiling against such a version?

Added one more commit to check for OpenSSL 1.1.1 or later during configure.

ifranzki commented:
I do see one testcase failure in ec_funcs with the soft token when running with OpenSSL 3.0:

`* TESTCASE run_ImportECCKeyPairSignVerify FAIL (testcases/crypto/ec_func.c:1530) C_CreateObject (EC Private Key) failed at i=19 (curve25519), rc=CKR_FUNCTION_FAILED`

This the first test for curve curve25519, and the soft token does not support Edwards/Montgomery curves jet. However, the testcases first imports the keys, and then check if the mechanism is supported by the token.  So I guess this is not an error in the token code, but in the testcase, it should not even try to import such keys if its not supported by the token. 

Nevertheless, I will check why it now fails (and did not fail with OpenSSL 1.1.1). I guess it has to do with the way how an PKEY key is constructed when importing an EC key. This is now much different with OpenSSL 3.0 than it was with OpenSSL 1.1.1.
Will update this PR when I know more.

p-steuer commented:
approved (if travis is green)

ifranzki commented:
So, Travis shows errors:

- x86: Problems with installing trousers. Not related to this PR I guess. Nevertheless, we should somehow fix this, but I don't really know how. Maybe by disabling the TPM token? Otherwise we will never see test runs on x86.....
- ppc64: `configure: error: OpenSSL 1.1.1 or later is required but OpenSSL libraries version 1.1.1 or later couldn't be found` . Thats somehow expected. We only support OpenSSL 1.1.1 or later now with this PR. Can we make sure that it uses OpenSSL 1.1.1 on ppc64, too ?
- clang (s390 and arm): Minor compile waring in p11sak tool. Not related to this PR, but I will look into fixing it.

ifranzki commented:
Rebased to current master (no change of the commits as such), and added 2 commits to fix the EC tests showing an error with OpenSSL 3.0 (error in testcase only), and the clang warning in p11sak (not related to OpenSSL 3.0 at all).

ifranzki commented:
Travis still shows the same errors for x86 and ppc64 (as expected), but s390 and arm are passed, for both, gcc and clang.

ifranzki commented:
Found a better fix for 
`* TESTCASE run_ImportECCKeyPairSignVerify FAIL (testcases/crypto/ec_func.c:1530) C_CreateObject (EC Private Key) failed at i=19 (curve25519), rc=CKR_FUNCTION_FAILED`
The OpenSSL 3.0 code did not return CKR_CURVE_NOT_SUPPORTED when an unsupported curve is used,
but with OpenSSL 1.1.1 it does return CKR_CURVE_NOT_SUPPORTED  as expected.
Fixed this now by returning CKR_CURVE_NOT_SUPPORTED in the OpenSSL 3.0 case, too. With this, the change in the EC testcase is no longer required, since it already checks for CKR_CURVE_NOT_SUPPORTED and skips the test. 

This was a small change in commit "SOFT: Remove deprecated OpenSSL functions" only.

ifranzki commented:
Travis still fails for ppc64 with not finding OpenSSL 1.1.1. 
I can't see from the travis output if it now runs on 20.04. The "Build system information" section of the output talks about "16.04".... But the job config shows 'dist: focal', but the 'apt-get install' takes packages from xenial ....

p-steuer commented:
Hmm, lets give PR https://github.com/opencryptoki/opencryptoki/pull/434 a try ..
